drivers/sensor/paw3212/paw3212.c

```diff
@@ -0,0 +1,609 @@
+}
+
+static int paw3212_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct paw3212_data *dev_data = &paw3212_data;
+	u8_t motion_status;
+	int err;
+
+	ARG_UNUSED(dev);
+
+	if (unlikely(!dev_data->ready)) {
+		LOG_INF("Device is not initialized yet");
+		return -EBUSY;
+	}
+
+	err = reg_read(dev_data, PAW3212_REG_MOTION, &motion_status);
+	if (err) {
+		LOG_ERR("Cannot read motion");
+		return err;
+	}
+
+	if ((motion_status & PAW3212_MOTION_STATUS_MOTION) != 0) {
+		u8_t x_low;
+		u8_t y_low;
+
+		if ((motion_status & PAW3212_MOTION_STATUS_DXOVF) != 0) {
+			LOG_WRN("X delta overlowed");
+		}
+		if ((motion_status & PAW3212_MOTION_STATUS_DYOVF) != 0) {
+			LOG_WRN("Y delta overlowed");
+		}
+
+		err = reg_read(dev_data, PAW3212_REG_DELTA_X_LOW, &x_low);
+		if (err) {
+			LOG_ERR("Cannot read X delta");
+			return err;
+		}
+
+		err = reg_read(dev_data, PAW3212_REG_DELTA_Y_LOW, &y_low);
+		if (err) {
+			LOG_ERR("Cannot read Y delta");
+			return err;
+		}
+
+		if (IS_ENABLED(CONFIG_PAW3212_12_BIT_MODE)) {
+			u8_t xy_high;
+
+			err = reg_read(dev_data, PAW3212_REG_DELTA_XY_HIGH,
+				       &xy_high);
+			if (err) {
+				LOG_ERR("Cannot read XY delta high");
+				return err;
+			}
+
+			dev_data->x = PAW3212_DELTA_X(xy_high, x_low);
+			dev_data->y = PAW3212_DELTA_Y(xy_high, y_low);
+		} else {
+			dev_data->x = (s8_t)x_low;
+			dev_data->y = (s8_t)y_low;
+		}
+	} else {
+		dev_data->x = 0;
+		dev_data->y = 0;
+	}
+
+	return err;
+}
+
+static int paw3212_channel_get(struct device *dev, enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	struct paw3212_data *dev_data = &paw3212_data;
+
+	ARG_UNUSED(dev);
+
+	if (unlikely(!dev_data->ready)) {
+		LOG_INF("Device is not initialized yet");
+		return -EBUSY;
+	}
+
+	switch (chan) {
+	case SENSOR_CHAN_POS_DX:
+		val->val1 = dev_data->x;
+		val->val2 = 0;
+		break;
+
+	case SENSOR_CHAN_POS_DY:
+		val->val1 = dev_data->y;
+		val->val2 = 0;
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int paw3212_trigger_set(struct device *dev,
+			       const struct sensor_trigger *trig,
+			       sensor_trigger_handler_t handler)
+{
+	struct paw3212_data *dev_data = &paw3212_data;
+	int err;
+
+	ARG_UNUSED(dev);
+
+	if (unlikely(trig->type != SENSOR_TRIG_DATA_READY)) {
+		return -ENOTSUP;
+	}
+
+	if (unlikely(trig->chan != SENSOR_CHAN_ALL)) {
+		return -ENOTSUP;
+	}
+
+	if (unlikely(!dev_data->ready)) {
+		LOG_INF("Device is not initialized yet");
+		return -EBUSY;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&dev_data->lock);
+
+	if (handler) {
+		err = gpio_pin_enable_callback(dev_data->irq_gpio_dev,
+					       CONFIG_PAW3212_IRQ_GPIO_PIN);
+	} else {
+		err = gpio_pin_disable_callback(dev_data->irq_gpio_dev,
+						CONFIG_PAW3212_IRQ_GPIO_PIN);
+	}
+
+	if (!err) {
+		dev_data->data_ready_handler = handler;
+	}
+
+	k_spin_unlock(&dev_data->lock, key);
+
+	return err;
+}
+
+static int paw3212_attr_set(struct device *dev, enum sensor_channel chan,
```

Comment:
We need to add attr set. I just don't want to add it in this PR as we have to have working FW for desktop mouse by the end of this week.

Reply:
If you need to add this function later, it would be better to add it then as a whole. I see no benefit of including now such useless piece of code. With `.attr_set = NULL` the driver will just look cleaner.